http-body/src/frame.rs

```diff
@@ -0,0 +1,68 @@
+use http::HeaderMap;
+
+/// A frame of any kind related to an HTTP stream (body).
+#[derive(Debug)]
+pub struct Frame<T> {
+    kind: Kind<T>,
+}
+
+#[derive(Debug)]
+enum Kind<T> {
+    // The first two variants are "inlined" since they are undoubtedly
+    // the most common. This saves us from having to allocate a
+    // boxed trait object for them.
+    Data(T),
+    Trailers(HeaderMap),
+    //Unknown(Box<dyn Frameish>),
+}
+
+impl<T> Frame<T> {
+    /// Create a DATA frame with the provided `Buf`.
+    pub fn data(buf: T) -> Self {
+        Self {
+            kind: Kind::Data(buf),
+        }
+    }
+
+    /// Create a trailers frame.
+    pub fn trailers(map: HeaderMap) -> Self {
+        Self {
+            kind: Kind::Trailers(map),
+        }
+    }
+
+    /// Returns whether this is a DATA frame.
+    pub fn is_data(&self) -> bool {
+        matches!(self.kind, Kind::Data(..))
+    }
+
+    /// Returns whether this is a trailers frame.
+    pub fn is_trailers(&self) -> bool {
+        matches!(self.kind, Kind::Trailers(..))
+    }
+
+    /// Consumes self into the buf of the DATA frame.
+    ///
+    /// # Panics
+    ///
+    /// If this isn't a DATA frame, this will panic. You should check `is_data`
+    /// before calling this.
+    // === REVIEW: is this the right method to expose? ===
```

Comment:
Is an `into_data() -> T` method that panics better, or `try_into_data() -> Result<T, Frame<T>>`?

Reply:
Why not have `into_data() -> Option<D>`? I feel like option gives the user to the choice of panic or result etc. With the option to get a ref as well? I feel like APIs that return a generic in the Err portion of result become really hard to use.